### PR TITLE
Parse variable feature syntax

### PR DESCRIPTION
**Pull request by cmyr**
This is a checkpoint; we can now parse the new syntax, but we do not compile anything. It's possible that some of this code will need to change as I dig into the compilation side, but I think it's worth getting this in and working from here.

### CODE REVIEW

File: fea-rs/src/parse/grammar/metrics.rs [hunk 1]
```diff
@@ -183,14 +248,23 @@ mod tests {
...
     }
 
+    #[test]
+    fn variable_anchor() {
+        let fea = "<anchor 0 (wght=200:-100 wght=900:-150 wght=900,wdth=150:-120)>";
+        let (_out, errors, _errstr) = debug_parse_output(fea, |parser| {
+            anchor(parser, TokenSet::EMPTY);
+        });
+        assert!(errors.is_empty());
+    }
```
Comment (rsheeter) [inline fea-rs/src/parse/grammar/metrics.rs:266]: The volume of changes to eat this, eat that, etc vs the volume of tests is making me skeptical. Are there not more variants to test?

**Review by rsheeter — APPROVED** [review]
LGTM with a side of skepticism about the volume of code vs test changes feeling off. Perhpas that is covered by the parse-tests updates, I'm not terribly familiar with the test strategy here.

File: fea-rs/src/parse/grammar/metrics.rs (hunk 1 above)
Comment (cmyr) [reply fea-rs/src/parse/grammar/metrics.rs:266]: I don't write unit tests for this; there's a system of property tests where we take some syntax and then compare the generated parse tree to an expected output. These are in the PR here: https://github.com/cmyr/fea-rs-test-data/pull/21

File: fea-rs/src/compile/compile_ctx.rs [hunk 2]
```diff
@@ -1695,34 +1712,40 @@ impl<'a> CompilationCtx<'a> {
...
+            }
+        } else if let Some((x_coord, y_coord)) = item.devices() {
+            Some(AnchorTable::format_3(
+                x,
+                y,
+                x_coord.compile(),
+                y_coord.compile(),
+            ))
+        } else {
+            Some(AnchorTable::format_1(x, y))
```
[the remark this answers is not in the document]
Comment (cmyr) [reply fea-rs/src/compile/compile_ctx.rs:1747]: This is actually after parsing, doing the final compilation pass on a validated AST. The 'resolve_' methods all take an AST node and figure out it actually represents.

I've pushed https://github.com/cmyr/fea-rs/pull/152 which adds a bit more overview-type docs to this file.

**Comment by cmyr** [comment]
To explain the testing strategy for parsing:

- We have two directories, `test-data/parsing/good` and `test-data/parsing/bad`
- both directories contain a bunch of feature files (`$file.fea`) each of which has a matching output file in the same directory.
- for the 'good' tests, the output file is `$file.PARSE_TREE`, which is a text file containing a human-readable dump of the AST.
- for the 'bad' tests, the output is `$file.ERR`, which contains a dump of the generated error message.
- there are a pair of tests in `src/tests/parse.rs`, one for the good and one for the bad cases.
- these tests parse every example in the two directories, ensuring that the output or errors exactly match what is expected.

This lets us ensure that our parsing behaviour is constant, and also that we are generating appropriate diagnostics when we encounter errors.


In addition to these parse tests we also have a similar system for testing compilation, where the 'expected output' is TTX.

**Comment by rsheeter** [comment]
> To explain the testing strategy for parsing:

Optional, IMHO this would be nice to capture, similar to how you explain achitecture etc on root https://github.com/cmyr/fea-rs

**Comment by cmyr** [comment]
It's documented in the readme of the `parse-tests` directory:

https://github.com/cmyr/fea-rs-test-data/tree/main/parse-tests

**Pull request merged**